File: src/main/java/org/taktik/icure/logic/FormTemplateLogic.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.logic;

import org.taktik.icure.dto.gui.layout.FormLayout;
import org.taktik.icure.entities.FormTemplate;

import java.util.List;

public interface FormTemplateLogic extends EntityPersister<FormTemplate, String> {
    FormTemplate get(String formId);
    List<FormTemplate> getFormTemplatesBySpecialty(String specialityCode, boolean loadLayout);
    List<FormTemplate> getFormTemplatesByUser(String id, boolean loadLayout);
    FormTemplate modifyFormTemplate(FormTemplate formTemplate);

	List<FormTemplate> getFormTemplatesByGuid(String userId, String specialityCode, String formTemplateGuid);

	FormLayout extractLayout(FormTemplate formTemplate);
    FormLayout build(byte[] data);
	List<String> getFieldsNames(FormLayout formTemplateDto);
	FormTemplate createFormTemplate(FormTemplate entity);
	FormTemplate getFormTemplateById(String formTemplateId);
}

File: src/main/kotlin/org/taktik/icure/asynclogic/FormTemplateLogic.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */

package org.taktik.icure.asynclogic

import kotlinx.coroutines.flow.Flow
import org.taktik.icure.asyncdao.FormTemplateDAO
import org.taktik.icure.dto.gui.layout.FormLayout
import org.taktik.icure.entities.FormTemplate

interface FormTemplateLogic : EntityPersister<FormTemplate, String> {
    fun createEntities(entities: Collection<FormTemplate>, createdEntities: Collection<FormTemplate>): Flow<FormTemplate>

    suspend fun createFormTemplate(entity: FormTemplate): FormTemplate

    suspend fun getFormTemplateById(formTemplateId: String): FormTemplate?
    fun getFormTemplatesByGuid(userId: String, specialityCode: String, formTemplateGuid: String): Flow<FormTemplate>
    fun getFormTemplatesBySpecialty(specialityCode: String, loadLayout: Boolean): Flow<FormTemplate>
    fun getFormTemplatesByUser(userId: String, loadLayout: Boolean): Flow<FormTemplate>

    suspend fun modifyFormTemplate(formTemplate: FormTemplate): FormTemplate?

    suspend fun build(data: ByteArray): FormLayout
    fun getGenericDAO(): FormTemplateDAO
    fun getFieldsNames(formLayout: FormLayout): List<String>
}
